docs/proposals/20210222-kubelet-authentication.md

```diff
@@ -0,0 +1,331 @@
+- The Kubernetes CSR spec has the needed groups
+- The Kubernetes CSR spec is limited to needed usages (e.g. client auth)
+- The Kubernetes CSR spec is limited to needed extensions (e.g. no CA extension)
+- Parse the CSR and verify that the CN is the same as .spec.username
+- Parse the CSR and verify that the Organization is the same as .spec.Groups
+- Parse the CSR and ensure that no SANs are appended for kubelet client certificates
+
+#### CSR format used by kubelet-authenticator
+We propose the introduction of X.509 extension attributes based
+on those reserved for the Kubernetes GCP cloud provider within Google’s organization ID allocation.
+
+We will request via SIG Architecture or CNCF to apply for an IANA OID registration block for the Kubernetes project.
+
+| OID Suffix | Name                                | Description                                                                                                                                                                                                                            |
+| ---------- | ----------------------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| 2.1.21     | KubernetesNodeProviderIdentifierOID | An identifier for the machine, should be the same or a derivative of the node provider ID. This is the equivalent of Google’s CloudComputeInstanceIdentifierOID, which we can reuse for a proof of concept (1.3.6.1.4.1.11129.2.1.21). |
+
+#### CSR PEM Blocks
+
+
+
+| Block Name                                 | Description                                          |
+| ------------------------------------------ | ---------------------------------------------------- |
+| KUBELET AUTHENTICATOR ATTESTATION PROVIDER | string describing the attestation provider           |
+| KUBELET AUTHENTICATOR ATTESTATION DATA     | the actual attestation data to perform validation on |
+
+#### Attestation data
+
+Attestation data will be appended, with any relevant base64 encoded data wrapped with the following header & footer.
+
+Example CSR:
+```
+-----BEGIN ATTESTATION DATA-----
+S25vd2luZyBtZSBBbGFuIFBhcnRyaWRnZSwga25vd2luZyB5b3UgS3ViZXJuZXRlcyBjbHVzdGVyLCBhaGEh
+-----END ATTESTATION DATA-----
+```
+The format of the attestation block is left to the provider. The Appendix section provides examples.
+
+#### Kubelet serving certificate rotation daemon
+
+In the first attempt, we will attempt to run serving certificate rotation as part of the client authenticator - it will
+check for the existence of the serving certificate files, their validity (e.g. not self signed), or request a new serving certificate. Once the new serving certificate is written, kubelet can be restarted.
+
+We may however find that it is better suited to run as a windows service manager service on the node.
```

Comment:
This statement seems random. I'd assume this is an exception for Windows?

Reply:
Yes